uaclient/cli.py

```diff
@@ -1619,6 +1620,64 @@ def action_refresh(args, *, cfg: config.UAConfig):
     return 0
 
 
+def is_contract_changed(cfg: config.UAConfig) -> bool:
+    orig_token = cfg.machine_token
+    orig_entitlements = cfg.entitlements
+    machine_token = orig_token["machineToken"]
+    contract_id = orig_token["machineTokenInfo"]["contractInfo"]["id"]
+    contract_client = contract.UAContractClient(cfg)
+    resp = contract_client.get_updated_contract_info(
+        machine_token, contract_id
+    )
+    try:
+        resp_ = json.dumps(resp, cls=util.DatetimeAwareJSONEncoder)
+        resp = json.loads(resp_, cls=util.DatetimeAwareJSONDecoder)
+    except ValueError:
+        return False
+    if (
+        cfg.contract_expiry_datetime
+        != resp["machineTokenInfo"]["contractInfo"]["effectiveTo"]
+    ):
+        return True
+    curr_entitlements = get_entitlements_from_response(resp)
+    for name, new_entitlement in sorted(curr_entitlements.items()):
+        deltas = util.get_dict_deltas(
+            orig_entitlements.get(name, {}), new_entitlement
+        )
+        if bool(deltas):
+            return True
+    return False
+
+
+def get_entitlements_from_response(data: Dict) -> Dict:
+    machine_token = data
+    if not machine_token:
+        return {}
+
+    entitlements = {}
+    contractInfo = machine_token.get("machineTokenInfo", {}).get(
+        "contractInfo"
+    )
+    if not contractInfo:
+        return {}
+
+    tokens_by_name = dict(
+        (e.get("type"), e.get("token"))
+        for e in machine_token.get("resourceTokens", [])
+    )
+    ent_by_name = dict(
+        (e.get("type"), e)
+        for e in contractInfo.get("resourceEntitlements", [])
+    )
+    for entitlement_name, ent_value in ent_by_name.items():
+        entitlement_cfg = {"entitlement": ent_value}
+        if entitlement_name in tokens_by_name:
+            entitlement_cfg["resourceToken"] = tokens_by_name[entitlement_name]
+        util.apply_contract_overrides(entitlement_cfg)
+        entitlements[entitlement_name] = entitlement_cfg
+    return entitlements
```

Comment:
This code is exactly what we do to process entitlements on the `config` module. Maybe we could modify the config module to receive a `machine_token` response. If we can do that, we would not need to recreate this function here, we would just compare the current `cfg.entitlements` with the `new_cfg.entitlements`.

What do you think @orndorffgrant @renanrodrigo ?

Reply:
Also had put it in my review - there are indeed a couple ways to reuse the logic, and I'd prefer that. Maybe some optional `from_machine_token` arg would be nice, yes

Reply:
True, it felt weird repeating the logic. Let me see how the suggestion will work out.